horovod/common/ops/adasum_cuda_ring_operations.cc

```diff
@@ -0,0 +1,198 @@
+
+  MPI_Comm_rank(mpi_context_->local_comm, &local_rank);
+
+  size_t layerid = 0;
+  //enqueue messages
+  for (; layerid < num_reductions; ++layerid) {
+    auto& entry = entries.at(layerid);
+    void* buffer_data;
+    int buffer_len;
+    void* recv_buffer;
+    buffer_data = (void*) entry.tensor->data();
+    buffer_len = entry.output->size();
+    if (layerid % 8 == local_rank) {
+      total_buffer_size += buffer_len;
+    }
+    if (entry.tensor->data() == entry.output->data()) {
+      // Get the temp buffer to be used for the Op
+      recv_buffer = (uint8_t*)global_buffer_data + offset;
+      offset += buffer_len;
+    }
+    else {
+      recv_buffer = (void*) entry.output->data();
+    }
+    gpu_context_->ErrorCheck("cudaSetDevice", cudaSetDevice(entry.device));
+
+    all_rings.InitMessageInRing(new ReduceMessage(mpi_context_),
+                      buffer_data,
+                      recv_buffer,
+                      buffer_len,
+                      entry.tensor->dtype(),
+                      mpi_context_->local_comm,
+                      layerid,
+                      global_state_->controller->GetLocalRank());
+  }
+  // wait for messages to finish
+  all_rings.WaitAllMessages();
+
+  if (cross_node) {
+      if (global_state_->parameter_manager.TensorFusionThresholdBytes() > total_buffer_size) {
+        total_buffer_size = global_state_->parameter_manager.TensorFusionThresholdBytes();
+      }
+      static void* allreduce_buffer = malloc(total_buffer_size);
+      static size_t cur_buffer_size = total_buffer_size;
+      if (total_buffer_size > cur_buffer_size) {
+        allreduce_buffer = realloc(allreduce_buffer, total_buffer_size);
+        cur_buffer_size = total_buffer_size;
+      }
+      offset = 0;
+      std::vector<int> tensor_counts(num_reductions);
+      // start device to host copies
+      for (layerid = 0; layerid < num_reductions; ++layerid) {
+        if (layerid % 8 == local_rank) {
+          auto& entry = entries.at(layerid);
+          tensor_counts[layerid] = entry.tensor->shape().num_elements();
+          int buffer_len = entry.output->size();
+          void* buffer_data = (uint8_t*)allreduce_buffer + offset;
+          offset += buffer_len;
+          
+          auto cuda_result = cudaMemcpyAsync(
+            buffer_data, (void*) entry.tensor->data(),
+            buffer_len, 
+            cudaMemcpyDeviceToHost,
+            gpu_context_->streams[global_state_->current_nccl_stream][layerid]);
+          gpu_context_->ErrorCheck("cudaMemcpyAsync", cuda_result);
+        } else {
+          tensor_counts[layerid] = 0;
+        }
+      }
+      auto recv_buffer = GetRecvBuffer(total_buffer_size);
+      // wait for this layer to finish copying to host
+      DispatchFusedAllreduce(entries, allreduce_buffer, recv_buffer, tensor_counts, local_size, 
+                              MPI_COMM_WORLD, 0, reduction_comms_,
```

Comment:
Rather than comm_world we should us the comm that horovod was initialized with.

Reply:
I noticed we were using MPI_COMM_WORLD for this in the hierarchical NCCL + Adasum code (adasum_nccl_operations.cc line 253). Was there no good reason for that, should I change this there as well?